File: lib/JFFStoreKit/JFFStoreKit/asyncAdapters/asyncSKPendingTransactions.m
#import "asyncSKPendingTransactions.h"

#import "JFFStoreKitDisabledError.h"

static NSString *const mergeObject = @"c8e5abce-1ab9-11e3-9a3b-f23c91aec05e";

@interface JFFAsyncSKPendingTransactions : NSObject <
SKPaymentTransactionObserver,
JFFAsyncOperationInterface
>
@end

@implementation JFFAsyncSKPendingTransactions
{
    SKPaymentQueue *_queue;
    BOOL _addedToObservers;
    JFFDidFinishAsyncOperationCallback _finishCallback;
}

- (void)dealloc
{
    [self unsubscribeFromObservervation];
    _finishCallback = nil;
}

- (void)doNothing:(id)objetc
{
}

- (void)unsubscribeFromObservervation
{
    if (_addedToObservers) {
        [_queue removeTransactionObserver:self];
        _addedToObservers = NO;
    }
}

+ (instancetype)newAsyncSKPendingTransactions
{
    JFFAsyncSKPendingTransactions *result = [self new];
    
    if (result) {
        result->_queue = [SKPaymentQueue defaultQueue];
        
        [result->_queue addTransactionObserver:result];
        result->_addedToObservers = YES;
    }
    
    return result;
}

- (void)asyncOperationWithResultCallback:(JFFDidFinishAsyncOperationCallback)finishCallback
                         handlerCallback:(JFFAsyncOperationChangeStateCallback)handlerCallback
                        progressCallback:(JFFAsyncOperationProgressCallback)progressCallback
{
    if (![SKPaymentQueue canMakePayments]) {
        finishCallback(nil, [JFFStoreKitDisabledError new]);
        return;
    }
    
    _finishCallback = [finishCallback copy];
    
    [[SKPaymentQueue defaultQueue] restoreCompletedTransactions];
}

- (void)doTask:(JFFAsyncOperationHandlerTask)task
{
    NSParameterAssert(task <= JFFAsyncOperationHandlerTaskCancel);
    
    if (task == JFFAsyncOperationHandlerTaskUnSubscribe)
        [self unsubscribeFromObservervation];
}

- (void)finishWithTransactions:(NSArray *)transactions
{
    transactions = [self restoredTransactionsForTransactions:transactions]?:@[];
    
    if (_finishCallback)
        _finishCallback(transactions, nil);
}

- (NSArray *)restoredTransactionsForTransactions:(NSArray *)transactions
{
    NSArray *result = [transactions select:^BOOL(SKPaymentTransaction *transaction) {
        return transaction.transactionState == SKPaymentTransactionStateRestored;
    }];
    
    return result;
}

#pragma mark SKPaymentTransactionObserver

- (void)paymentQueue:(SKPaymentQueue *)queue updatedTransactions:(NSArray *)transactions
{
}

- (void)paymentQueueRestoreCompletedTransactionsFinished:(SKPaymentQueue *)queue
{
    JFFAsyncSKPendingTransactions *self_ = self;
    
    [self_ finishWithTransactions:queue.transactions];
    [self_ unsubscribeFromObservervation];
}

- (void)paymentQueue:(SKPaymentQueue *)queue restoreCompletedTransactionsFailedWithError:(NSError *)error
{
    JFFAsyncSKPendingTransactions *self_ = self;
    
    if (_finishCallback)
        _finishCallback(nil, error);
    
    [self_ unsubscribeFromObservervation];
}

@end

static JFFAsyncOperation privateAllPendingTransactionsLoader()
{
    JFFAsyncOperationInstanceBuilder factory = ^id< JFFAsyncOperationInterface >() {
        return [JFFAsyncSKPendingTransactions newAsyncSKPendingTransactions];
    };
    
    JFFAsyncOperation loader = buildAsyncOperationWithAdapterFactory(factory);
    
    return [mergeObject asyncOperationMergeLoaders:loader withArgument:@(__FUNCTION__)];
}

JFFAsyncOperation allPendingTransactionsLoader()
{
    JFFAsyncOperation loader = privateAllPendingTransactionsLoader();
    
    return loader;
}

File: iAsync.Swift/lib/JStoreKit/JStoreKit/asyncAdapters/asyncSKPendingTransactions.swift
//
//  asyncSKPendingTransactions.swift
//  JStoreKit
//
//  Created by Vladimir Gorbenko on 03.10.14.
//  Copyright (c) 2014 EmbeddedSources. All rights reserved.
//

import Foundation

import JAsync
import JUtils
import StoreKit

private let cachedAsyncOp = JCachedAsync<String, [SKPaymentTransaction]>()

//TODO private
public class JAsyncSKPendingTransactions : NSObject, SKPaymentTransactionObserver, JAsyncInterface {
    
    public typealias ResultType = [SKPaymentTransaction]
    
    private let queue: SKPaymentQueue
    private var addedToObservers: Bool
    private var finishCallback: JAsyncTypes<ResultType>.JDidFinishAsyncCallback?
    
    deinit {
        unsubscribeFromObservervation()
        finishCallback = nil
    }
    
    @objc(doNothing:)//TODO remove
    public func doNothing(objetc: AnyObject) {}
    
    func unsubscribeFromObservervation() {
        if addedToObservers {
            queue.removeTransactionObserver(self)
            addedToObservers = false
        }
    }
    
    override init() {
        
        queue = SKPaymentQueue.defaultQueue()
        addedToObservers = true
        
        super.init()
        
        queue.addTransactionObserver(self)
    }
    
    public func asyncWithResultCallback(
        finishCallback: JAsyncTypes<ResultType>.JDidFinishAsyncCallback,
        stateCallback: JAsyncChangeStateCallback,
        progressCallback: JAsyncProgressCallback)
    {
        if !SKPaymentQueue.canMakePayments() {
            finishCallback(result: JResult.error(JStoreKitDisabledError()))
            return
        }
    
        self.finishCallback = finishCallback
    
        SKPaymentQueue.defaultQueue().restoreCompletedTransactions()
    }
    
    public func doTask(task: JAsyncHandlerTask) {
        
        assert(task.rawValue <= JAsyncHandlerTask.Cancel.rawValue)
    
        if task == .UnSubscribe {
            unsubscribeFromObservervation()
        }
    }
    
    public var isForeignThreadResultCallback: Bool {
        return false
    }
    
    func finishWithTransactions(transactions: [SKPaymentTransaction]?) {
    
        let restoredTransactions = restoredTransactionsForTransactions(transactions ?? [])
        
        self.finishCallback?(result: JResult.value(restoredTransactions))
    }
    
    public func paymentQueue(queue: SKPaymentQueue!, updatedTransactions transactions: [AnyObject]!) {}
    
    func restoredTransactionsForTransactions(transactions: [SKPaymentTransaction]) -> [SKPaymentTransaction] {
    
        let result = transactions.filter { (transaction: SKPaymentTransaction) -> Bool in
            return transaction.transactionState == SKPaymentTransactionState.Restored
        }
    
        return result
    }

    public func paymentQueueRestoreCompletedTransactionsFinished(queue: SKPaymentQueue!) {
        
        let self_ = self
        
        self_.finishWithTransactions(queue.transactions as? [SKPaymentTransaction])
        self_.unsubscribeFromObservervation()
    }
    
    public func paymentQueue(queue: SKPaymentQueue!, restoreCompletedTransactionsFailedWithError error: NSError!) {
        
        let self_ = self

        finishCallback?(result: JResult.error(error))

        self_.unsubscribeFromObservervation()
    }
}

public func allPendingTransactionsLoader() -> JAsyncTypes<[SKPaymentTransaction]>.JAsync {
    
    let factory = { () -> JAsyncSKPendingTransactions in
        return JAsyncSKPendingTransactions()
    }
    
    let loader = JAsyncBuilder.buildWithAdapterFactory(factory)
    
    return cachedAsyncOp.asyncOpMerger(loader, uniqueKey:__FUNCTION__)
}
